<font size=4><b>Coursera Capstone Project Notebook</b></font>

This notebook will be used for the completion of the Coursera Capstone Project in Python.

In [1]:
import numpy as np
import pandas as pd

Need the following installs in order to complete this project

In [2]:
from geopy.geocoders import Nominatim #for letting location data
from sklearn.cluster import KMeans #for performing kmeans clustering
import folium #for visualizing on a world map

In [3]:
print('Hello Capstone Project Course!')

Hello Capstone Project Course!


Load and view our dataset

In [4]:
data = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]

In [5]:
data.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

We need to remove all of the not assigned buroughs from our dataset.

In [6]:
data = data[data['Borough'] != 'Not assigned']

In [7]:
data.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights

We need to combine rows with the same postcode

In [8]:
neighborhoods_in_pc = {}
for postcode in data['Postcode'].unique():
    pc_data = data[data['Postcode'] == postcode]
    neighborhoods_in_pc[postcode] = pc_data['Neighbourhood'].unique()
neighborhoods_in_pc

{'M3A': array(['Parkwoods'], dtype=object),
 'M4A': array(['Victoria Village'], dtype=object),
 'M5A': array(['Harbourfront', 'Regent Park'], dtype=object),
 'M6A': array(['Lawrence Heights', 'Lawrence Manor'], dtype=object),
 'M7A': array(['Not assigned'], dtype=object),
 'M9A': array(['Islington Avenue'], dtype=object),
 'M1B': array(['Rouge', 'Malvern'], dtype=object),
 'M3B': array(['Don Mills North'], dtype=object),
 'M4B': array(['Woodbine Gardens', 'Parkview Hill'], dtype=object),
 'M5B': array(['Ryerson', 'Garden District'], dtype=object),
 'M6B': array(['Glencairn'], dtype=object),
 'M9B': array(['Cloverdale', 'Islington', 'Martin Grove', 'Princess Gardens',
        'West Deane Park'], dtype=object),
 'M1C': array(['Highland Creek', 'Rouge Hill', 'Port Union'], dtype=object),
 'M3C': array(['Flemingdon Park', 'Don Mills South'], dtype=object),
 'M4C': array(['Woodbine Heights'], dtype=object),
 'M5C': array(['St. James Town'], dtype=object),
 'M6C': array(['Humewood-Cedarvale'

Create a new dataframe with postcodes having values of every neighborhood in that postcode

In [9]:
fixed_data = pd.DataFrame(columns=['Postcode','Borough','Neighbourhood'])
fixed_data['Postcode'] = neighborhoods_in_pc.keys()
fixed_data['Neighbourhood'] = neighborhoods_in_pc.values()
fixed_data.head()

Postcode Borough                       Neighbourhood
0      M3A     NaN                         [Parkwoods]
1      M4A     NaN                  [Victoria Village]
2      M5A     NaN         [Harbourfront, Regent Park]
3      M6A     NaN  [Lawrence Heights, Lawrence Manor]
4      M7A     NaN                      [Not assigned]

Now, we need to find the Burough for each Postcode

In [10]:
buroughs = []
for pc in fixed_data['Postcode']:
    pc_data = data[data['Postcode'] == pc] #only postcard data in here
    buroughs.append(pc_data['Borough'].unique())
fixed_data['Borough'] = buroughs

In [11]:
fixed_data.head()

Postcode             Borough                       Neighbourhood
0      M3A        [North York]                         [Parkwoods]
1      M4A        [North York]                  [Victoria Village]
2      M5A  [Downtown Toronto]         [Harbourfront, Regent Park]
3      M6A        [North York]  [Lawrence Heights, Lawrence Manor]
4      M7A      [Queen's Park]                      [Not assigned]

Columns without assigned Neighbourhood, Neighbourhood should be set to Borough

In [12]:
for index, row in fixed_data.iterrows():
    if (row['Neighbourhood'][0] == 'Not assigned'):
        row['Neighbourhood'][0] = row['Borough']

In [13]:
fixed_data.head(25)

Postcode             Borough  \
0       M3A        [North York]   
1       M4A        [North York]   
2       M5A  [Downtown Toronto]   
3       M6A        [North York]   
4       M7A      [Queen's Park]   
5       M9A         [Etobicoke]   
6       M1B       [Scarborough]   
7       M3B        [North York]   
8       M4B         [East York]   
9       M5B  [Downtown Toronto]   
10      M6B        [North York]   
11      M9B         [Etobicoke]   
12      M1C       [Scarborough]   
13      M3C        [North York]   
14      M4C         [East York]   
15      M5C  [Downtown Toronto]   
16      M6C              [York]   
17      M9C         [Etobicoke]   
18      M1E       [Scarborough]   
19      M4E      [East Toronto]   
20      M5E  [Downtown Toronto]   
21      M6E              [York]   
22      M1G       [Scarborough]   
23      M4G         [East York]   
24      M5G  [Downtown Toronto]   

                                        Neighbourhood  
0                                         [Parkwoods]  
1                                  [Victoria Village]  
2                         [Harbourfront, Regent Park]  
3                  [Lawrence Heights, Lawrence Manor]  
4                                    [[Queen's Park]]  
5                                  [Islington Avenue]  
6                                    [Rouge, Malvern]  
7                                   [Don Mills North]  
8                   [Woodbine Gardens, Parkview Hill]  
9                          [Ryerson, Garden District]  
10                                        [Glencairn]  
11  [Cloverdale, Islington, Martin Grove, Princess...  
12           [Highland Creek, Rouge Hill, Port Union]  
13                 [Flemingdon Park, Don Mills South]  
14                                 [Woodbine Heights]  
15                                   [St. James Town]  
16                               [Humewood-Cedarvale]  
17  [Bloordale Gardens, Eringate, Markland Wood, O...  
18                [Guildwood, Morningside, West Hill]  
19                                      [The Beaches]  
20                                      [Berczy Park]  
21                              [Caledonia-Fairbanks]  
22                                           [Woburn]  
23                                          [Leaside]  
24                               [Central Bay Street]

Clean up the Borough column

In [14]:
fixed_boroughs = []
fixed_neighbourhoods = []
for borough in fixed_data['Borough']:
    b = ','.join(borough)
    fixed_boroughs.append(b)
fixed_data['Borough'] = fixed_boroughs

Clean up the Neighbourhood column

In [15]:
fixed_neighbourhoods = []
for neighborhood in fixed_data['Neighbourhood']:
    if len(neighborhood) == 1:
        n = neighborhood[0]
    else:
        n = ",".join(str(x) for x in neighborhood)
    fixed_neighbourhoods.append(n)
fixed_data['Neighbourhood'] = fixed_neighbourhoods

In [16]:
fixed_data.head(10)

Postcode           Borough                    Neighbourhood
0      M3A        North York                        Parkwoods
1      M4A        North York                 Victoria Village
2      M5A  Downtown Toronto         Harbourfront,Regent Park
3      M6A        North York  Lawrence Heights,Lawrence Manor
4      M7A      Queen's Park                   [Queen's Park]
5      M9A         Etobicoke                 Islington Avenue
6      M1B       Scarborough                    Rouge,Malvern
7      M3B        North York                  Don Mills North
8      M4B         East York   Woodbine Gardens,Parkview Hill
9      M5B  Downtown Toronto          Ryerson,Garden District

Assumptions: I am assuming that each postcode can contain multiple neighbourhoods and does not contain any overlapping neighbourhoods. I am also assuming each borough falls into 1 postcode (and each postcode has 1 Borough).

In [17]:
fixed_data.shape

(103, 3)

103 rows and 3 columns.

Now, we need to get the longitude and latitude using GeoPy

In [18]:
neighborhoods = fixed_data['Neighbourhood']
latitudes = []
longitudes = []
for city in neighborhoods:
    try:
        user_agent_name = city.replace(' ','_') + '_explorer'
        geolocator = Nominatim(user_agent=user_agent_name)
        location = geolocator.geocode(city)
        latitude = location.latitude
        longitude = location.longitude
        latitudes.append(latitude)
        longitudes.append(longitude)
        print('The geograpical coordinate of ', city, ' are {}, {}.'.format(latitude, longitude))
    except:
        try:
            user_agent_name = city.replace(' ','_') + '_explorer'
            geolocator = Nominatim(user_agent=user_agent_name)
            location = geolocator.geocode(city)
            latitude = location.latitude
            longitude = location.longitude
            latitudes.append(latitude)
            longitudes.append(longitude)
            print('The geograpical coordinate of ', city, ' are {}, {}.'.format(latitude, longitude))
        except:
            latitudes.append(np.NaN)
            longitudes.append(np.NaN)

The geograpical coordinate of  Parkwoods  are 26.5652643, -81.8817227102639.
The geograpical coordinate of  Victoria Village  are 43.732658, -79.3111892.
The geograpical coordinate of  Harbourfront,Regent Park  are 43.6400801, -79.3801495.
The geograpical coordinate of  Lawrence Heights,Lawrence Manor  are 43.7227784, -79.4509332.
The geograpical coordinate of  Islington Avenue  are 43.6393743, -79.5212175.
The geograpical coordinate of  Rouge,Malvern  are 43.8091955, -79.2217008.
The geograpical coordinate of  Don Mills North  are 43.737178, -79.3434514.
The geograpical coordinate of  Ryerson,Garden District  are 45.5797934, -79.5084658621578.
The geograpical coordinate of  Glencairn  are -34.1595402, 18.4284987.
The geograpical coordinate of  Woodbine Heights  are 43.6999302, -79.3191316.
The geograpical coordinate of  St. James Town  are 43.6694032, -79.3727041.
The geograpical coordinate of  Humewood-Cedarvale  are 43.69079835, -79.4253981936993.
The geograpical coordinate of  Guil

In [19]:
fixed_data['Latitude'] = latitudes
fixed_data['Longitude'] = longitudes
fixed_data.dropna(inplace=True,how='any')

In [20]:
fixed_data.head(10)

Postcode           Borough                    Neighbourhood   Latitude  \
0       M3A        North York                        Parkwoods  26.565264   
1       M4A        North York                 Victoria Village  43.732658   
2       M5A  Downtown Toronto         Harbourfront,Regent Park  43.640080   
3       M6A        North York  Lawrence Heights,Lawrence Manor  43.722778   
5       M9A         Etobicoke                 Islington Avenue  43.639374   
6       M1B       Scarborough                    Rouge,Malvern  43.809196   
7       M3B        North York                  Don Mills North  43.737178   
9       M5B  Downtown Toronto          Ryerson,Garden District  45.579793   
10      M6B        North York                        Glencairn -34.159540   
14      M4C         East York                 Woodbine Heights  43.699930   

    Longitude  
0  -81.881723  
1  -79.311189  
2  -79.380150  
3  -79.450933  
5  -79.521218  
6  -79.221701  
7  -79.343451  
9  -79.508466  
10  18.428499  
14 -79.319132

Now, we need to do some visualizations

In [40]:
geo_map = folium.Map(
    location=[43.5,-80],
    zoom_start=8,
    tiles='Stamen Terrain'
)

In [41]:
geo_map

Let's add markers to our map

Make a new df containing only unique rows

In [42]:
marker_df = fixed_data.drop_duplicates(subset='Neighbourhood',keep='first',inplace=False)

In [43]:
for index, row in marker_df.iterrows():
    city = folium.map.FeatureGroup()
    city.add_child(
        folium.CircleMarker(
        [row['Latitude'],row['Longitude']], radius=5,
        color='red',
        fill_color='Red'
        )
    )
    geo_map.add_child(city)
    folium.Marker([row['Latitude'],row['Longitude']],popup=row['Neighbourhood']).add_to(geo_map)

In [44]:
geo_map